server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderIT.java

```
@@ -128,29 +133,80 @@ public void testRestoreSnapshotAllocationDoesNotExceedWatermark() throws Excepti
+            ClusterInfoServiceUtils.refresh(clusterInfoService);
+            shardRelocationsTracker.process(event);
+        });
+
+        final String dataNode0Id = internalCluster().getInstance(NodeEnvironment.class, dataNodeName).nodeId();
+
+        final String indexName = randomIdentifier();
+        createIndex(indexName, indexSettings(6, 0).put(INDEX_STORE_STATS_REFRESH_INTERVAL_SETTING.getKey(), "0ms").build());
+        var shardSizes = createReasonableSizedShards(indexName);
+
+        final CreateSnapshotResponse createSnapshotResponse = clusterAdmin().prepareCreateSnapshot("repo", "snap")
+            .setWaitForCompletion(true)
+            .get();
+        final SnapshotInfo snapshotInfo = createSnapshotResponse.getSnapshotInfo();
+        assertThat(snapshotInfo.successfulShards(), is(snapshotInfo.totalShards()));
+        assertThat(snapshotInfo.state(), is(SnapshotState.SUCCESS));
+
+        assertAcked(indicesAdmin().prepareDelete(indexName).get());
+        updateClusterSettings(Settings.builder().put(CLUSTER_ROUTING_REBALANCE_ENABLE_SETTING.getKey(), Rebalance.NONE.toString()));
+        shardRelocationsTracker.reset();
+
+        // reduce disk size of node 0 so that only a single shard can be allocated
+        getTestFileStore(dataNodeName).setTotalSpace(shardSizes.getBiggestShardSize() + WATERMARK_BYTES + 1L);
+        refreshDiskUsage();
+
+        final RestoreSnapshotResponse restoreSnapshotResponse = clusterAdmin().prepareRestoreSnapshot("repo", "snap")
+            .setWaitForCompletion(true)
+            .get();
+        final RestoreInfo restoreInfo = restoreSnapshotResponse.getRestoreInfo();
+        assertThat(restoreInfo.successfulShards(), is(snapshotInfo.totalShards()));
+        assertThat(restoreInfo.failedShards(), is(0));
+
+        assertBusyWithDiskUsageRefresh(dataNode0Id, indexName, new ContainsExactlyOneOf<>(shardSizes.getAllShardIds()));
+        shardRelocationsTracker.assertRelocationCount(0);
```

Could we fail the test as soon as we see an unexpected relocation, rather than here at the end?

---

I did it this way to avoid failing on another thread and to make the test cleaner by placing all the assertions in the end

---

Understood, but I don't see a problem with failing on another thread and I think it's easier to debug failures exactly when they happen.

---

Sounds good, updated